File: 49.group-anagrams.java
/*
 * @lc app=leetcode id=49 lang=java
 *
 * [49] Group Anagrams
 */

// @lc code=start

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        for (String str : strs) {
            char[] chars = str.toCharArray();
            Arrays.sort(chars);
            String keystr = new String(chars);
            map.putIfAbsent(keystr, new ArrayList<>());
            map.get(keystr).add(str);
        }
        return new ArrayList<>(map.values());
    }
}
// @lc code=end

File: 49.group-anagrams.cpp
/*
 * @lc app=leetcode id=49 lang=cpp
 *
 * [49] Group Anagrams
 */

// @lc code=start

#include <vector>
#include <string>
#include <map>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        map<string, vector<string>> m;
        for_each(strs.begin(), strs.end(), [&m](const string s) {
            vector<char> chars;
            copy(s.begin(), s.end(), back_inserter(chars));
            sort(chars.begin(), chars.end());
            string key {chars.begin(), chars.end()};
            m[key].push_back(s);
        });
        vector<vector<string>> ret;
        for_each(m.begin(), m.end(), [&ret](const auto p) {
            ret.push_back(p.second);
        });
        return ret;
    }
};
// @lc code=end
